### ISSUE TITLE
Removing vertical halos for `ShallowWaterModel`

### ISSUE THREAD

**Issue by francispoulin** [CliMA/Oceananigans.jl#1499]
Now that `ShallowWaterModel` is established I think it's important to make it more efficient.

At the moment when we use `WENO5`, as I've been doing, we define 3 halo points in each direction.  That means that our 1-layer model is really a 7-layer model, which is much more storage than we need.  I very much want to remove this so that we can be much more efficient and have the code run faster.

I know that #1024 tried to establish a flat dimension for `IncompressibleModel`.  This is before my time and I heard multiple people tried doing this and it was a challenge. 

I would like to do something independent of #1024, and force that any grid for `ShallowWaterModel` not define halo regions in the vertical and then not fill in those halo regions as it evolves.  Below are a list of different things that I think should be done but am very curious to know what people suggest.  I haven't tried this so I'm sure there are some details that I have not thought of and hoping to learn from others.

- [ ] Maybe define the boundary conditions to be `flat` in the vertical and then try and get flat to work for this model?
- [ ] Maybe in `field_boundary_conditions.jl`, define `uhFluxBoundaryCondtions`, `vhFluxBoundaryCondtions`, `hBoundaryCondtions` that do not do anything to `top` and `bottom`?
- [ ] If we set the number of halos in the vertical to be 0, then we should still be able to use `new_data.jl` without having to modify it?
- [ ] Tell `fill_halo_regions.jl` to not do anything for `bottom` and `top`.
- [ ] If we remove the halos then I don't think we have to modify the tendencies at all so that might be fine.
- [ ] Other things that I've not considered?

**Comment by navidcy** [comment]
That's a good idea. I can't comment on the technicalities. 

But this should be the case for all flat dimensions. 

Perhaps we should finish off https://github.com/CliMA/Oceananigans.jl/pull/1024 and then do this in general for all Flat dimensions.

**Comment by francispoulin** [comment]
Thanks @navidcy .  I just strated a new branch that will try and do this for shallow water and maybe the other models, ignoring all the closure functions.  `fjp/flat-for-shallow-water-model`.  Nothings yet but will play with it and let people know how it develops.

**Comment by francispoulin** [comment]
I have a start to this soon to be PR (I hope) but nothing working  yet.  

One issue that comes up is that `automatic_halo_sizing.jl` should not extend the halo if the topology is flat  This is easy enough to set this up in principle, however, when I try telling it what `Flat` means using the following `using Oceananigans.Grids: Flat`, it fails.  

How should I load `Flat` at this stage?

**Comment by glwagner** [comment]
The code for setting up `RegularRectilinearGrid` with a `Flat` dimension should be in place.

The issues being resolved on #1024 are associated with finite volume operations; for example if the grid spacing is 0 in the vertical, then the rectilinear volume is zero. This leads to 0/0 and thus NaNs for the operators used in `IncompressibleModel`.

For `ShallowWaterModel` the situation is different because different operators are used. Presumably there are fewer issues with `ShallowWaterModel`.

@francispoulin can you please try to use `Flat` in one of the shallow water examples and tell us what error you get, or if you get NaNs?

**Comment by glwagner** [comment]
For example, change these lines:

https://github.com/CliMA/Oceananigans.jl/blob/80b0aad76cc4ca70cc146177ca19e34ff790d51c/examples/shallow_water_Bickley_jet.jl#L37-L39

to

```julia
 grid = RegularRectilinearGrid(size = (Nx, Ny, 1), 
                             x = (0, Lx), y = (-Ly/2, Ly/2), z = (0, Lz), 
                             topology = (Periodic, Bounded, Flat)) 
```

Note there is an alignment issue in that example.

**Comment by francispoulin** [comment]
Thanks @glwagner and I have been using #1024 a whole lot in developing `fjp/flat-for-shallow-water-model`.  So far the example above works for the grid part and the model part.  When I run a solution I get `NaN` after the first time step when one of the fields is `Flat`.  I suspect this is because the derivatives might need to be modified to account for a `Flat` dimension.  I hope to know more tomorrow.

**Comment by francispoulin** [comment]
It seems that one needs to be careful on how we define the areas and volumes, as @glwagner mentioned previously.  In particular, in `spacings_and_arrays_and_volumes.jl`, there are the following two lines that appear currently in master, and I believe these are correct without `Flat` in the vertical,

```
@inline ΔzF(i, j, k, grid::RegularRectilinearGrid) = grid.Δz
@inline ΔzF(i, j, k, grid::VerticallyStretchedRectilinearGrid) = @inbounds grid.Δzᵃᵃᶜ[k]
```

However, in #1024, it is suggested that if `Flat` in the vertical then we need to pick the vertical spacing to be unity, as such

```
@inline ΔzF(i, j, k, grid::AbstractGrid{FT, TX, TY, Flat}) where {FT, TX, TY} = one(FT)
```

Unfortunately, I don't know how to properly dispatch between these different cases as there is overlap between `RegularRectilinearGrid` and `Flat` in the vertical.  Any suggestions?

**Comment by glwagner** [comment]
> as there is overlap between RegularRectilinearGrid and Flat in the vertical.

What do you mean?

**Comment by francispoulin** [comment]
The test cases using a RegularRectilinearGrid and Flat.  The dispatch seems to use the first one but I think we need to define the dispatch a bit differently, but I don't know how to do that.  A subdispatch sounds attractive but don't know if that's really a thing.

**Comment by glwagner** [comment]
Ah sorry I still don't understand what you're saying. Can you elaborate? 

Note that when `topology = (Periodic, Periodic, Flat)` (for example), then we should have `grid <: AbstractGrid{FT, TX, TY, Flat} where {FT, TX, TY}` (the fourth type parameter is the topology in the vertical).

https://github.com/CliMA/Oceananigans.jl/blob/80b0aad76cc4ca70cc146177ca19e34ff790d51c/src/Grids/regular_rectilinear_grid.jl#L8

**Comment by francispoulin** [comment]
Sorry for not being clear, wil try again with code.  Consider the following.

```
    grid = RegularRectilinearGrid(size=(4,4), extent=(1,1), topology=(Periodic,Periodic,Flat))
    model = ShallowWaterModel(grid=grid, gravitational_acceleration=1)
```

The first line says we are using a `RegularRectilinearGrid` and the second line says we have `Flat` in the vertical.  

The three options to compute the vertical grid spacing are the following,

```
@inline ΔzF(i, j, k, grid::RegularRectilinearGrid) = grid.Δz
@inline ΔzF(i, j, k, grid::VerticallyStretchedRectilinearGrid) = @inbounds grid.Δzᵃᵃᶜ[k]
@inline ΔzF(i, j, k, grid::AbstractGrid{FT, TX, TY, Flat}) where {FT, TX, TY} = one(FT)
```
It seems like the first and third functions are both options.  When I tried it, even though I had `Flat` in the vertical, it used the first and computed the spacing in the vertical to be `0` instead of `1`.  This resulted in a lot of `NaNs`.

**Comment by glwagner** [comment]
Can you try

```julia
@inline ΔzF(i, j, k, grid::AbstractGrid{FT, TX, TY, <:Flat}) where {FT, TX, TY} = one(FT)
```

?

There may be some small subtle bug here.

**Comment by francispoulin** [comment]
thanks for the suggestion.  I tried it and it still seems to use the first function in the list above.  

I guess I'm not sure how the grids are defined differently in the different cases.

I should say that the third method above was taken was taken from #1024, and I thought was designed to do this, but that might have been before `VerticallyStretchedRectilinearGrid` matured.

**Comment by glwagner** [comment]
No, the method looks correct to me. This is some kind of bug.

As a sanity check, you might try defining

```julia
@inline ΔzF(i, j, k, grid::RegularRectilinearGrid{FT, TX, TY, Flat}) where {FT, TX, TY} = one(FT)
```

However, we definitely want to get the method working for `AbstractGrid` so that we don't have to redefine `ΔzF` for every grid type.

**Comment by glwagner** [comment]
You can also try something like

```julia
julia> const VerticallyFlatGrid = AbstractGrid{FT, TX, TY, <:Flat} where {FT, TX, TY}

julia> typeof(grid) <: VerticallyFlatGrid
```

I hope the answer is `true`!

**Comment by glwagner** [comment]
> I guess I'm not sure how the grids are defined differently in the different cases.

What do you mean?

**Comment by francispoulin** [comment]
> ```julia
> typeof(grid) <: VerticallyFlatGrid
> ```

Happy to say it is true!

**Comment by francispoulin** [comment]
> > I guess I'm not sure how the grids are defined differently in the different cases.
> 
> What do you mean?

In the three methods we seem to have three different types of grids were considering,

```
RegularRectilinearGrid
VerticallyStretchedRectilinearGrid
AbstractGrid{FT, TX, TY, Flat}
```

From the above, it is definitely `RegularRectilinearGrid` with a `Flat` dimension  I looked at `Grids.jl` and have learned that `RegularRectilinearGrid` is of type `AbstractRectilinearGrid` and this is a subtype of `AbstractGrid.  I guess this explains the hiearchy.

**Comment by francispoulin** [comment]
If I use `@inline ΔzF(i, j, k, grid::AbstractGrid{FT, TX, TY, Flat}) where {FT, TX, TY} = one(FT)` then I get that for the 2D case with `Flat` only in the vertical, than all the tendencies work out correctly.  However, when I try a 1D case with `Flat` in both `y` and `z`, things go the way of `NaNs` again.  

So figuring out how to do this dispatch correctly is a big first step, then I have to be more careful of how to define the areas for the 1D and 0D case.  The same strategy, whatever it is we find, should work, for the other problem.

**Comment by francispoulin** [comment]
Problem solved!  @ali-ramadhan figured out that if we instead use the following then julia figures out how to dispatch correctly

```
@inline Δx(i, j, k,  grid::RegularRectilinearGrid{FT, Flat})         where FT           = one(FT)
@inline Δy(i, j, k,  grid::RegularRectilinearGrid{FT, TX, Flat})     where {FT, TX}     = one(FT)
@inline ΔzC(i, j, k, grid::RegularRectilinearGrid{FT, TX, TY, Flat}) where {FT, TX, TY} = one(FT)
@inline ΔzF(i, j, k, grid::RegularRectilinearGrid{FT, TX, TY, Flat}) where {FT, TX, TY} = one(FT)
```

I will do some tests, cleaning up and then push a PR to move the disucssion there.

**Comment by glwagner** [comment]
Are we happy with this solution? It's slightly boiler-platy. `Flat` won't be supported automatically when we add new grids. We really do probably want this for `CubedSphere` shallow water models.

Maybe we need to specify that a grid is _not_ `Flat` for the fallback implementations?

```julia
const PeriodicOrBounded = Union{Periodic, Bounded}
const NonFlatRegularRectilinearGrid = RegularRectilinearGrid{FT, <:PeriodicOrBounded, <:PeriodicOrBounded, <:PeriodicOrBounded} where FT
```

etc?

**Comment by glwagner** [comment]
> Happy to say it is true!

I'm totally perplexed why dispatch seems to be failing silently. Often in these cases we get an error that a method is ambiguous. I'm stumped.

**Comment by glwagner** [comment]
> From the above, it is definitely RegularRectilinearGrid with a Flat dimension I looked at Grids.jl and have learned that RegularRectilinearGrid is of type AbstractRectilinearGrid and this is a subtype of `AbstractGrid. I guess this explains the hiearchy.

Yes, that is the type hierarchy and why I thought the above code pattern would work.

**Comment by francispoulin** [comment]
> Are we happy with this solution? It's slightly boiler-platy. `Flat` won't be supported automatically when we add new grids. We really do probably want this for `CubedSphere` shallow water models.
> 
> Maybe we need to specify that a grid is _not_ `Flat` for the fallback implementations?
> 
> ```julia
> const PeriodicOrBounded = Union{Periodic, Bounded}
> const NonFlatRegularRectilinearGrid = RegularRectilinearGrid{FT, <:PeriodicOrBounded, <:PeriodicOrBounded, <:PeriodicOrBounded} where FT
> ```
> 
> etc?

I agree that what I just pushed probably is not the ideal in terms of good programming style.  It works but I think we can do better.  What you suggest seems to make sense to me.  What do others think is the best way forward?

Also, since I just created a PR, do we want to move this discussion there?

**Comment by ali-ramadhan** [comment]
@francispoulin Should this issue be closed now that PR #1504 has been merged?

**Comment by francispoulin** [comment]
Yes, and thanks for not forgetting about it.
